### PR TITLE
Max stamina modifiers from mutations now affect stamina regen

### PR DESCRIPTION
**Pull request by CodeBandit**
#### Summary
```SUMMARY: Bugfixes "Max stamina modifiers from mutations now affect stamina regen"```

#### Purpose of change
To normalize the time needed to recover stamina to 500 turns regardless of maximum stamina, and make certain traits make more sense in terms of time to recover stamina. Previously, languorous only needed 375 turns to recover all max stamina, and indefatigable required 625. Realistically, it makes no sense for an "unfit" person to recover faster than a "fit" person.

This also affects walking stamina regeneration, where indefatigable would gain 10 stamina instead of 5, and languorous would be at 0 net stamina from walking. The numbers for walking emulate fit/unfit stamina more realistically this way. Sprinting and crouching are barely affected.

#### Describe the solution
Apply max_stamina_modifier to stamina_multiplier for regen purposes. Movement costs remain untouched.

Stamina regen are as follows:
no mutation: 20 (net 5 stamina gain when walking)
Languorous: 15 (net 0 stamina loss when walking)
indefatigable: 25 (net 10 stamina gain when walking)

#### Testing
Enable debug mode to see stamina consumption
Check stamina regen and consumption across no mutations, languorous, and indefatigable.

#### Notes
This is a rebase of #39533 . 
Open question: Is there an easier way to rebase on git without all the commit verbosity? (Answered by anothersimulacrum in a different PR)

### CODE REVIEW

**Comment by Tamiore** [comment]
Feels like `languorous` and `indefatigable` trait points costs should be modified.
Doubled, I guess?

**Comment by CodeBandit** [comment]
@Tamiore Please see #39533 since it is a re-PR. It was previously discussed, and it looks like Kevin is okay with it so long as Languorous didn't enter the negative net stamina regen territory.

**Comment by Tamiore** [comment]
I don't think Kevin ever commented directly on the point cost of the traits. 

I mean, indefatigable at 2 points is ok-ish. Definitely among the best traits now, but not completely without precedent. 

But languorous at -2 is really, reeeally punishing.

**Comment by CodeBandit** [comment]
> I don't think Kevin ever commented directly on the point cost of the traits.

That's true, I would also like to know his stance on the trait costs assuming this gets merged.

> I mean, indefatigable at 2 points is ok-ish. Definitely among the best traits now, but not completely without precedent.

Could I ask you to elaborate on that?

> But languorous at -2 is really, reeeally punishing. Arguably to the point few players will take it.
Few players take languorous already. Those who do only take it as a challenge/RP trait. Arguably, it makes it more attractive. If min-max is the perspective, I feel min-max is, unfortunately, a symptom of a broken system (assigning points to traits -personal opinion, and definitely tangential to the topic).

**Comment by Tamiore** [comment]
> Could I ask you to elaborate on that?

What other 2-pointers are also extremely good? 
Parkour expert and Fleet-footed, to name a couple.

**Comment by CodeBandit** [comment]
I'll phrase is differently, how would this make Indefatigable be among the best traits?

**Comment by Tamiore** [comment]
> I'll phrase is differently, how would this make Indefatigable be among the best traits?

Because nothing can harm you if nothing can catch you. 

And right now the only zombies that a quick + fleet-footed character can't consistently outrun are feral hunters/predators. 

And the reason you can't outrun them is because you can't consistently put enough distance to prevent them from catching up to you. 

With Indefatigable, however, not only will you be able to run 25% further away, but you will also run *faster*. (Because your stamina % depletes slower, so speed penalties kick in later.)

That alone was reason enough to consider taking Indefatigable.  

Now — with extra 25% stamina regeneration —Indefatigable extend you running even further (afaik, you still regenerate stamina when you run, and extra 5/turn adds up to a few extra % of effective speed/distance). 

Not only that, but it increases your speed at full stamina, too. (Better stamina regen = more surplus stamina when walking = you can use it to run for a tile every 8 moves or so.)

**Comment by CodeBandit** [comment]
> With Indefatigable, however, not only will you be able to run 25% further away, but you will also run _faster_. (Because your stamina % depletes slower, so speed penalties kick in later.)

_faster_ is incorrect, running farther doesn't mean you run faster. If you can't break line of sight with something faster than you, or you can't kill it, or you can't get away from it -how much max stamina you have is inconsequential.

> Now — with extra 25% stamina regeneration —Indefatigable extend your running even further (afaik, you still regenerate stamina when you run, and extra 5/turn adds up to a few extra % of effective speed/distance).

Lets do the math for absolute best case scenario (stamina % has no effect):
running is 7x base stamina burn rate (which is 15), so that's a cost of 105.
before indefatigable:
10000/(105 - 20) = 117.6 tiles
Indefatigable, before change:
12500/(105 - 20) = 147.1 tiles (that's 25% increase over trait-less as expected)
Indefatigable, after change:
12500/(105 - 25) = 156.25 tiles

That's 9 tiles more in the absolute _best_ case. It would be much _much_ less than that if stamina drain was taken into consideration. In this case, it's not even worthwhile since, if you're running out your stamina anyways, you're dead.

> Not only that, but it increases your speed at full stamina, too. (Better stamina regen = more surplus stamina when walking = you can use it to run for a tile every 8 moves or so.)

That's not how _speed_ works.

> TL;DR: with "Quick + Fleet-Footed" Indefatigable is exactly the final boost you need to outrun effectively ANY zombie.

I agree it pairs well with the speed traits, but don't lump in things with things that make you faster 
since this doesn't.

**Comment by Tamiore** [comment]
>faster is incorrect, running farther doesn't mean you run faster.

Again, since your stamina % drops slower with Indefatigable, you do run effectively faster as well.
It doesn't effect your *maximum* speed, but it does raise your *effective* running speed if you run long enough. Because with larger stamina pool your speed *penalty* builds up slower. 

By the time your stamina would drop to 50% without Indefatigable, you will only be down to 60% stamina with Indefatigable. And at 60% stamina your *speed* penalty is lower than at 50% stamina.

>If you can't break line of sight with something faster than you, or you can't kill it, or you can't get away from it -how much max stamina you have is inconsequential.

No zombie is "faster than you" when you start running from full stamina with fleet-footed and quick. At that point it is **exactly** a question of how far you can get before you run out of stamina.
Do you want me to record you a video that shows how you can't run away from feral predator with Indefatigable, but **can** with Indefatigable?

>That's 9 tiles more in the absolute best case.

And that's a HUGE difference when you are trying to break line of sight against fast Z's in the actual game.

>That's not how speed works.

Do we really need to argue over semantics here? The result is that you can cover ground quicker while maintaining full stamina.
Does it really matter that **technically** it wasn't your *speed* that changed, but *proportion of moves you can spend running while maintaining zero net stamina consumption*?

>I agree it pairs well with the speed traits, but don't lump in things with things that make you faster since this doesn't.
Yes, it DOES make you faster, if my "faster" you mean "how much ground you will cover with one X seconds".
See above why.

**Comment by CodeBandit** [comment]
The argument isn't for "with vs without" indefatigable, it's indefatigable "before vs after" the change. Largely, it _should_ be able whether or not stamina modifiers should affect stamina regeneration. Taking a step back before this conversation spirals out of focus, I feel, in light of the conversation, this doesn't bring Indefatigable or languorous outside the scope of a 2-point trait (personal opinion).

**Comment by Tamiore** [comment]
>The argument isn't for "with vs without" indefatigable, it's indefatigable "before vs after" the change.

Having extra stamina regeneration still somewhat slows stamina pool depletion rate while running. So the same logic still applies. The actual difference in effective speed is minor, but it's there. 

>Taking a step back before this conversation spirals out of focus, I feel, in light of the conversation, this doesn't bring Indefatigable or languorous outside the scope of a 2-point trait (personal opinion).

I feel that Indefatigable moved from being "ok" trait at 2 points to being "one of the best". It's not unreasonable at 2 points, just really good.

As for languorous, I still think -3 or -4 would be more appropriate.
